File: Framework/Util/Transform.lua
require('vector');
require('wrench');
require('twist');

----------------------------------------------------------------------
-- Transform : homogeneous transform
----------------------------------------------------------------------

Transform = {};
Transform.__index = Transform;
Transform.__mtstring = 'Transform';

-- Constructors
--------------------------------------------------------------------------

function Transform.eye()
  local t = {};
  t[1] = vector.new({1, 0, 0, 0});
  t[2] = vector.new({0, 1, 0, 0});
  t[3] = vector.new({0, 0, 1, 0});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.rotZ(a)
  local ca = math.cos(a);
  local sa = math.sin(a);
  local t = {};
  t[1] = vector.new({ca, -sa, 0, 0});
  t[2] = vector.new({sa, ca, 0, 0});
  t[3] = vector.new({0, 0, 1, 0});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.rotY(a)
  local ca = math.cos(a);
  local sa = math.sin(a);
  local t = {};
  t[1] = vector.new({ca, 0, sa, 0});
  t[2] = vector.new({0, 1, 0, 0});
  t[3] = vector.new({-sa, 0, ca, 0});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.rotX(a)
  local ca = math.cos(a);
  local sa = math.sin(a);
  local t = {};
  t[1] = vector.new({1, 0, 0, 0});
  t[2] = vector.new({0, ca, -sa, 0});
  t[3] = vector.new({0, sa, ca, 0});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.transX(dx)
  local t = {};
  t[1] = vector.new({1, 0, 0, dx});
  t[2] = vector.new({0, 1, 0, 0});
  t[3] = vector.new({0, 0, 1, 0});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.transY(dy)
  local t = {};
  t[1] = vector.new({1, 0, 0, 0});
  t[2] = vector.new({0, 1, 0, dy});
  t[3] = vector.new({0, 0, 1, 0});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.transZ(dz)
  local t = {};
  t[1] = vector.new({1, 0, 0, 0});
  t[2] = vector.new({0, 1, 0, 0});
  t[3] = vector.new({0, 0, 1, dz});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.rotation(r)
  local t = {}
  assert(#r == 9);
  t[1] = vector.new({r[1], r[2], r[3], 0});
  t[2] = vector.new({r[4], r[5], r[6], 0});
  t[3] = vector.new({r[7], r[8], r[9], 0});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.translation(p)
  local t = {};
  assert(#p == 3);
  t[1] = vector.new({1, 0, 0, p[1]});
  t[2] = vector.new({0, 1, 0, p[2]});
  t[3] = vector.new({0, 0, 1, p[3]});
  t[4] = vector.new({0, 0, 0, 1});
  return setmetatable(t, Transform);
end

function Transform.pose(p)
  --[[
  return transform :
  t = Transform.transX(p[1])
  t = Transform.transY(p[2])
  t = Transform.transZ(p[3])
  t = t*Transform.rotX(p[4])
  t = t*Transform.rotY(p[5])
  t = t*Transform.rotZ(p[6])
  --]]
  local t = Transform.eye(); 
  for i = 4,6 do
    p[i] = p[i] or 0;
  end
  assert(#p == 6);
  local cwx = math.cos(p[4]);
  local swx = math.sin(p[4]);
  local cwy = math.cos(p[5]);
  local swy = math.sin(p[5]);
  local cwz = math.cos(p[6]);
  local swz = math.sin(p[6]);
  t[1][1] = cwy*cwz; 
  t[1][2] = -cwy*swz;
  t[1][3] = swy; 
  t[1][4] = p[1];
  t[2][1] = cwx*swz + swx*swy*cwz;
  t[2][2] = cwx*cwz - swx*swy*swz;
  t[2][3] = -swx*cwy;
  t[2][4] = p[2]; 
  t[3][1] = swx*swz - cwx*swy*cwz;
  t[3][2] = swx*cwz + cwx*swy*swz; 
  t[3][3] = cwx*cwy; 
  t[3][4] = p[3]; 
  t[4][1] = 0; 
  t[4][2] = 0; 
  t[4][3] = 0; 
  t[4][4] = 1; 
  return setmetatable(t, Transform);
end

function Transform.euler(w)
  --[[
  return transform :
  t = t*Transform.rotX(w[1])
  t = t*Transform.rotY(w[2])
  t = t*Transform.rotZ(w[3])
  --]]
  assert(#w == 3);
  return Transform.pose({0, 0, 0, w[1], w[2], w[3]});
end

-- Methods
--------------------------------------------------------------------------

function Transform.inv(t)
  tinv = Transform.eye();
  for i = 1,3 do 
    for j = 1,3 do
      -- Transpose rotation:
      tinv[i][j] = t[j][i];
      -- Compute inv translation:
      tinv[i][4] = tinv[i][4] - t[j][i]*t[j][4];
    end
  end
  return tinv;
end

function Transform.get_rotation(t)
  local r = {};
  r[1] = t[1][1];
  r[2] = t[1][2];
  r[3] = t[1][3];
  r[4] = t[2][1];
  r[5] = t[2][2];
  r[6] = t[2][3];
  r[7] = t[3][1];
  r[8] = t[3][2];
  r[9] = t[3][3];
  return vector.new(r);
end

function Transform.get_translation(t)
  local p = {}; 
  p[1] = t[1][4];
  p[2] = t[2][4];
  p[3] = t[3][4];
  return vector.new(p);
end

function Transform.get_pose(t)
  -- returns 6 DOF translation and rotation {tx, ty, tz, rx, ry, rz}
  local p = vector.zeros(6);
  local w = Transform.get_euler(t);
  p[1] = t[1][4];
  p[2] = t[2][4];
  p[3] = t[3][4];
  p[4] = w[1];
  p[5] = w[2];
  p[6] = w[3];
  return p;
end

function Transform.get_euler(t)
  -- returns euler angles {wx, wy, wz} corresponding to rotation RxRyRz
  local w = vector.zeros(3);
  w[2] = math.asin(t[1][3]);
  if (t[1][3] == -1) then
    w[1] = -math.atan2(t[2][1], t[2][2]);
    w[3] = 0;
  elseif (t[1][3] == 1) then
    w[1] = math.atan2(t[2][1], t[2][2]);
    w[3] = 0;
  else
    w[1] = math.atan2(-t[2][3]/math.cos(w[2]), t[3][3]/math.cos(w[2])); 
    w[3] = math.atan2(-t[1][2]/math.cos(w[2]), t[1][1]/math.cos(w[2]));
  end
  return w;
end

function Transform.__tostring(t)
  local str = ''
  for i = 1,4 do
    for j = 1,4 do
      str = str..string.format('%.4f ', t[i][j]);
    end
    str = str..'\n';
  end
  return str;
end

function Transform.__mul(t1, t2)
  if (getmetatable(t2) == vector) then
    local v = {};
    for i = 1,4 do
      v[i] = t1[i][1] * t2[1]
              + t1[i][2] * t2[2]
              + t1[i][3] * t2[3]
              + t1[i][4] *(t2[4] or 1);
    end
    for i = 1,4 do
      v[i] = v[i]/v[4];
    end
    return setmetatable(v, vector);
  elseif (getmetatable(t2) == Transform) then
    local t = {};
    for i = 1,4 do
      t[i] = {};
      for j = 1,4 do
        t[i][j] = t1[i][1] * t2[1][j]
                  + t1[i][2] * t2[2][j]
                  + t1[i][3] * t2[3][j]
                  + t1[i][4] * t2[4][j];
      end
    end
    return setmetatable(t, Transform);
  elseif (getmetatable(t2) == wrench) then
    local M = Transform.rotation(t1:get_rotation());
    local p = t1:get_translation();
    local w = wrench.new();
    w:set_force(M*t2:get_force());
    w:set_torque(M*t2:get_torque());
    return w:translate(-p);
  elseif (getmetatable(t2) == twist) then
    local M = Transform.rotation(t1:get_rotation());
    local p = t1:get_translation();
    local t = twist.new();
    t:set_lin(M*t2:get_lin());
    t:set_rot(M*t2:get_rot());
    return t:translate(-p);
  else
    error('attempt to multiply Transform by incompatible type');
  end
end

return Transform;

File: thro/Platform/ash/Comms/Transform.cpp
#include "Transform.h"
#include <math.h>
#include <stdio.h>

Transform::Transform() {
  clear();
}

void Transform::clear() {
  // Initialize to identity matrix:
  for (int i = 0; i < 4; i++)
    for (int j = 0; j < 4; j++)
      t[i][j] = 0;

  t[0][0] = 1;
  t[1][1] = 1;
  t[2][2] = 1;
  t[3][3] = 1;
}

Transform& Transform::translate(double x, double y, double z) {
  t[0][3] += t[0][0]*x + t[0][1]*y + t[0][2]*z;
  t[1][3] += t[1][0]*x + t[1][1]*y + t[1][2]*z;
  t[2][3] += t[2][0]*x + t[2][1]*y + t[2][2]*z;
  return *this;
}

Transform& Transform::translateX(double x) {
  t[0][3] += t[0][0]*x;
  t[1][3] += t[1][0]*x;
  t[2][3] += t[2][0]*x;
  return *this;
}

Transform& Transform::translateY(double y) {
  t[0][3] += t[0][1]*y;
  t[1][3] += t[1][1]*y;
  t[2][3] += t[2][1]*y;
  return *this;
}

Transform& Transform::translateZ(double z) {
  t[0][3] += t[0][2]*z;
  t[1][3] += t[1][2]*z;
  t[2][3] += t[2][2]*z;
  return *this;
}

Transform& Transform::rotateX(double a) {
  double ca = cos(a);
  double sa = sin(a);
  for (int i = 0; i < 3; i++) {
    double ty = t[i][1];
    double tz = t[i][2];
    t[i][1] = ca*ty + sa*tz;
    t[i][2] = -sa*ty + ca*tz;
  }
  return *this;
}

Transform& Transform::rotateY(double a) {
  double ca = cos(a);
  double sa = sin(a);
  for (int i = 0; i < 3; i++) {
    double tx = t[i][0];
    double tz = t[i][2];
    t[i][0] = ca*tx - sa*tz;
    t[i][2] = sa*tx + ca*tz;
  }
  return *this;
}

Transform& Transform::rotateZ(double a) {
  double ca = cos(a);
  double sa = sin(a);
  for (int i = 0; i < 3; i++) {
    double tx = t[i][0];
    double ty = t[i][1];
    t[i][0] = ca*tx + sa*ty;
    t[i][1] = -sa*tx + ca*ty;
  }
  return *this;
}

Transform& Transform::mDH(double alpha, double a, double theta, double d) {
  /*
  Transform t1;
  double ca = cos(alpha);
  double sa = sin(alpha);
  double ct = cos(theta);
  double st = sin(theta);
  t1(0,0) = ct; t1(0,1) = -st; t1(0,2) = 0; t1(0,3) = a;
  t1(1,0) = st*ca; t1(1,1) = ct*ca; t1(1,2) = -sa; t1(1,3) = -sa*d;
  t1(2,0) = st*sa; t1(2,1) = ct*sa; t1(2,2) = ca; t1(2,3) = ca*d;
  */

  this->translateX(a).rotateX(alpha).translateZ(d).rotateZ(theta);
  return *this;
}

void Transform::apply(double x[3]) {
  double x0[3];
  for (int i = 0; i < 3; i++) {
    x0[i] = x[i];
  }
  for (int i = 0; i < 3; i++) {
    x[i] = t[i][3];
    for (int j = 0; j < 3; j++) {
      x[i] += t[i][j]*x0[j];
    }
  }
}

double const Transform::operator() (int i, int j) const {
  return t[i][j];
}

double& Transform::operator() (int i, int j) {
  return t[i][j];
}

Transform operator* (const Transform &t1, const Transform &t2) {
  Transform t;
  for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 4; j++) {
      t(i,j) = t1(i,0)*t2(0,j) + t1(i,1)*t2(1,j) +
	t1(i,2)*t2(2,j) + t1(i,3)*t2(3,j);
    }
  }
  return t;
}

Transform inv (const Transform &t1) {
  Transform t;
  for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 3; j++) {
      // Transpose rotation:
      t(i,j) = t1(j,i);
      // Compute inv translation:
      t(i,3) -= t1(j,i)*t1(j,3);
    }
  }
  return t;
}

Transform transform6D(const double p[6]) {
  Transform t;
  //  t = t.translate(p[0],p[1],p[2]).rotateX(p[3]).rotateY(p[4]).rotateZ(p[5]);
  double cwx = cos(p[3]);
  double swx = sin(p[3]);
  double cwy = cos(p[4]);
  double swy = sin(p[4]);
  double cwz = cos(p[5]);
  double swz = sin(p[5]);
  t(0,0) = cwy*cwz; 
  t(0,1) = -cwy*swz;
  t(0,2) = swy; 
  t(0,3) = p[0];
  t(1,0) = cwx*swz + swx*swy*cwz;
  t(1,1) = cwx*cwz - swx*swy*swz;
  t(1,2) = -swx*cwy;
  t(1,3) = p[1]; 
  t(2,0) = swx*swz - cwx*swy*cwz;
  t(2,1) = swx*cwz + cwx*swy*swz; 
  t(2,2) = cwx*cwy; 
  t(2,3) = p[2]; 
  t(3,0) = 0; 
  t(3,1) = 0; 
  t(3,2) = 0; 
  t(3,3) = 1; 
  return t;
}

std::vector<double> position6D(const Transform &t1) {
  std::vector<double> p(6);
  p[0] = t1(0,3);
  p[1] = t1(1,3);
  p[2] = t1(2,3);
  // calculate euler angles corresponding to rotation matrix RxRyRz
  p[4] = asin(t1(0,2));
  if (t1(0,2) == -1) {
    p[3] = -atan2(t1(1,0), t1(1,1));
    p[5] = 0;
  }
  else if (t1(0,2) == 1) {
    p[3] = atan2(t1(1,0), t1(1,1));
    p[5] = 0;
  }
  else {
    p[3] = atan2(-t1(1,2)/cos(p[4]), t1(2,2)/cos(p[4])); 
    p[5] = atan2(-t1(0,1)/cos(p[4]), t1(0,0)/cos(p[4]));
  }
  return p;
}

void printTransform(Transform tr) {
  for (int i = 0; i < 4; i++) {
    for (int j = 0; j < 4; j++) {
      printf("%.4g ", tr(i,j));
    }
    printf("\n");
  }
  printf("\n");
}

void printVector(std::vector<double> v) {
  for (int i = 0; i < v.size(); i++) {
    printf("%.4g\n", v[i]);
  }
  printf("\n");
}
